src/driver/safe/core.rs

```diff
@@ -559,68 +505,96 @@ unsafe impl Sync for CudaFunction {}
+        Ok(CudaStream {
+            cu_stream: result::stream::create(result::stream::StreamKind::NonBlocking)?,
+            forked_from: None,
+            device: self.clone(),
+        })
+    }
+
     /// Allocates a new stream that can execute kernels concurrently to the default stream.
     ///
     /// The synchronization with default stream happens in **code order**. See [CudaStream] docstring.
     ///
     /// This stream synchronizes in the following way:
     /// 1. On creation it adds a wait for any existing work on the default work stream to complete
     /// 2. On drop it adds a wait for any existign work on Self to complete *to the default stream*.
+    #[deprecated]
     pub fn fork_default_stream(self: &Arc<Self>) -> Result<CudaStream, result::DriverError> {
-        self.bind_to_thread()?;
-        let stream = CudaStream {
-            stream: result::stream::create(result::stream::StreamKind::NonBlocking)?,
-            device: self.clone(),
-        };
-        stream.wait_for_default()?;
-        Ok(stream)
+        self.default_stream().fork()
     }
 
     /// Forces [CudaStream] to drop, causing the default work stream to block on `stream`'s completion.
     /// **This is asynchronous with respect to the host.**
     #[allow(unused_variables)]
+    #[deprecated]
     pub fn wait_for(self: &Arc<Self>, stream: &CudaStream) -> Result<(), result::DriverError> {
-        self.bind_to_thread()?;
-        unsafe {
-            result::event::record(self.event, stream.stream)?;
-            result::stream::wait_event(
-                self.stream,
-                self.event,
-                sys::CUevent_wait_flags::CU_EVENT_WAIT_DEFAULT,
-            )
-        }
+        self.default_stream().wait_for(stream)
     }
 }
 
 impl CudaStream {
+    pub fn fork(&self) -> Result<CudaStream, result::DriverError> {
+        let mut stream = self.device.new_stream()?;
+        stream.forked_from = Some(self.cu_stream);
+        stream.wait_for(&self)?;
+        Ok(stream)
+    }
+
     /// Records the current default stream's workload, and then causes `self`
     /// to wait for the default stream to finish that recorded workload.
-    pub fn wait_for_default(&self) -> Result<(), result::DriverError> {
-        self.device.bind_to_thread()?;
+    pub fn wait_for(&self, other: &Self) -> Result<(), result::DriverError> {
+        let event = result::event::create(sys::CUevent_flags::CU_EVENT_DISABLE_TIMING)?;
         unsafe {
-            result::event::record(self.device.event, self.device.stream)?;
+            result::event::record(event, other.cu_stream)?;
             result::stream::wait_event(
-                self.stream,
-                self.device.event,
+                self.cu_stream,
+                event,
                 sys::CUevent_wait_flags::CU_EVENT_WAIT_DEFAULT,
-            )
+            )?;
+            // TODO test that this is safe
+            result::event::destroy(event)
```

Comment:
The cuda docs say
> An event may be destroyed before it is complete (i.e., while [cuEventQuery()](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EVENT.html#group__CUDA__EVENT_1g6f0704d755066b0ee705749ae911deef) would return [CUDA_ERROR_NOT_READY](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1ggc6c391505e117393cc2558fff6bfc2e9edd9cef666ce620352e619a36b6c3f34)). In this case, the call does not block on completion of the event, and any associated resources will automatically be released asynchronously at completion.

My interpretation of this is that we can destroy an event and it will still do the expected synchronization. Just need to test this